nutils/types.py

```diff
@@ -349,7 +349,6 @@ def __get__(self, instance, owner):
             cached_value = getattr(instance, self.cache_attr)
         except AttributeError:
             value = self.fget(instance)
-            assert _isimmutable(value)
```

Comment:
can you explain in the commit message why this condition can be relaxed? I don't immediately see it

Reply:
Clear. It seems to me that option 3 is doable if we recognize bound methods (using inspect.isinstance) and testing immutability of `__self__`, functions (for staticmethods) and numpy ufuncs as a third special category. What do you think?

Reply:
This seems to work: `isinstance(value, numpy.ufunc)`. I'll try that.

Reply:
We have decided to go for option 2 after all, as it appears to be difficult to cover all possible values of `evalf`, in particular instances of external classes that define a `__call__` method (e.g. `nutils_poly.MulPlan` which will be introduced in #700).

Reply:
A side effect of using `cached_property` is that we need a `__dict__`. Not a big problem IMO. We could even consider dropping `__slots__`.

Reply:
Agreed. I suspect there was never a strong reason to use those, and probably less so today with developments such as https://github.com/python/cpython/pull/28802

Reply:
I just noticed that currently you keep `__slots__` and add `__dict__`. Better drop `__slots__` I would say, even if we decide to keep the remaining ones for now.

Reply:
PS when we do ultimately drop all slots I think we can consider replacing all of `__cache__` by `cached_property` decorators.